DATA CLEANING AND PREPARATION FOR AVIATION ACCIDENT DATA
Below we are importing Pandas 

In [12]:
#importing pandas for data manipulation
import pandas as pd

Using "read_csv", a panda function, we were able to read the csv file to a form python can manipulate the data and saved under a DataFrame "df".

In [13]:
#read csv
df = pd.read_csv('aviation-accident-data-2023-05-16.csv')
df.head()

date             type registration       operator fatalities  \
0  date unk.   Antonov An-12B       T-1206  Indonesian AF        NaN   
1  date unk.   Antonov An-12B       T-1204  Indonesian AF        NaN   
2  date unk.   Antonov An-12B       T-1201  Indonesian AF        NaN   
3  date unk.  Antonov An-12BK          NaN      Soviet AF        NaN   
4  date unk.  Antonov An-12BP   CCCP-11815      Soviet AF          0   

              location          country cat     year  
0                  NaN  Unknown country  U1  unknown  
1                  NaN  Unknown country  U1  unknown  
2                  NaN  Unknown country  U1  unknown  
3  Tiksi Airport (IKS)           Russia  A1  unknown  
4  Massawa Airport ...          Eritrea  A1  unknown

After reading the csv file into a form it can be manipulated by python we are able to clean the data.
We then have to check the number of columns and their data types to see what we are working with.

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23967 entries, 0 to 23966
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   date          23967 non-null  object
 1   type          23967 non-null  object
 2   registration  22419 non-null  object
 3   operator      23963 non-null  object
 4   fatalities    20029 non-null  object
 5   location      23019 non-null  object
 6   country       23967 non-null  object
 7   cat           23967 non-null  object
 8   year          23967 non-null  object
dtypes: object(9)
memory usage: 1.6+ MB


From the above function we were able to see that there are 9 columns with all of them being object data type which needs to change.

Below we then checked which columns had missing values and how many they were.

In [15]:
df.isna().sum()

date               0
type               0
registration    1548
operator           4
fatalities      3938
location         948
country            0
cat                0
year               0
dtype: int64

Using "dropna" ,which is a method in python, we used it to remove the missing values from the DataFrame.
Using "inplace =True" we used it to affect the actual database otherwise nothing would've been affected.

In [16]:
df.dropna(inplace = True)

As you can see below we checked for the missing values and it brought null.

In [17]:
df.isna().sum()

date            0
type            0
registration    0
operator        0
fatalities      0
location        0
country         0
cat             0
year            0
dtype: int64

For the second step in the data cleaning process we checked for duplicated values.

In [18]:
df.duplicated().sum()

1

Only one was found.
We removed the duplicated value in the using the method "drop_duplicates" for data accuracy.

In [19]:
df.drop_duplicates(inplace = True)

As you can see below we checked again and duplicated values is null.

In [20]:
df.duplicated().sum()

0

In [21]:
df.head()

date             type registration    operator fatalities  \
4   date unk.  Antonov An-12BP   CCCP-11815   Soviet AF          0   
7   date unk.     Antonov An-2   CCCP-01216    Aeroflot          0   
9   date unk.    Antonov An-26       01 red   Soviet AF          0   
10  date unk.    Antonov An-26     RA-47415  Russian AF          0   
13  date unk.    Antonov An-2R            -  PZL-Mielec          0   

               location  country cat     year  
4   Massawa Airport ...  Eritrea  A1  unknown  
7          Chita region   Russia  A2  unknown  
9     Orenburg Air Base   Russia  O1  unknown  
10  Belgorod Airport...   Russia  A1  unknown  
13               Mielec   Poland  O1  unknown

Below we used the method (".apply") to only affect the column "fatalities" as our goal is to evaluate the column from object data type to integer data type as it would be easily manipulated and be able to calculated.

In [22]:
df['fatalities'] = df['fatalities'].apply(lambda x: eval(str(x)) if isinstance(x, str)else x)
#if isinstance (x,str) else x part is just a safety check to make sure we only eval to strings and leave actual numbers unchanged.

When we checked again below for the data types, the column "fatalities" datatype changed from object to int64

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18642 entries, 4 to 23966
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   date          18642 non-null  object
 1   type          18642 non-null  object
 2   registration  18642 non-null  object
 3   operator      18642 non-null  object
 4   fatalities    18642 non-null  int64 
 5   location      18642 non-null  object
 6   country       18642 non-null  object
 7   cat           18642 non-null  object
 8   year          18642 non-null  object
dtypes: int64(1), object(8)
memory usage: 1.4+ MB


In [24]:
df.head()

date             type registration    operator  fatalities  \
4   date unk.  Antonov An-12BP   CCCP-11815   Soviet AF           0   
7   date unk.     Antonov An-2   CCCP-01216    Aeroflot           0   
9   date unk.    Antonov An-26       01 red   Soviet AF           0   
10  date unk.    Antonov An-26     RA-47415  Russian AF           0   
13  date unk.    Antonov An-2R            -  PZL-Mielec           0   

               location  country cat     year  
4   Massawa Airport ...  Eritrea  A1  unknown  
7          Chita region   Russia  A2  unknown  
9     Orenburg Air Base   Russia  O1  unknown  
10  Belgorod Airport...   Russia  A1  unknown  
13               Mielec   Poland  O1  unknown

The next step we noticed that some values in the column "year" was stated "unknown" in which we used python to detect how many they were.

In [25]:
df['year'].eq('unknown').sum()

16

Using the method (".loc") we removed the rows containing "unknown".
All rows that DIDN'T state "unknown" was saved back to the DataFrame.
("!=" means NOT equal to)

In [26]:
df = df.loc[df['year'] != 'unknown']
df.head()

date                type registration                operator  \
62  11-AUG-1919     Felixstowe Fury         N123                     RAF   
63  23-FEB-1920    Handley Page O/7       G-EANV  Handley Page Transport   
64  25-FEB-1920  Handley Page O/400       G-EAMC  Handley Page Transport   
65  30-JUN-1920  Handley Page O/400       G-EAKE  Handley Page Transport   
66  14-DEC-1920  Handley Page O/400       G-EAMA  Handley Page Transport   

    fatalities             location       country cat  year  
62           1  near Felixtowe RNAS          U.K.  A1  1919  
63           0  Acadia Siding, C...  South Africa  A1  1920  
64           0      near El Shereik         Sudan  A1  1920  
65           0             Östanå        Sweden  A1  1920  
66           4        Golders Green          U.K.  A1  1920

For confirmation we checked again to see if there were "unknown" values and brought back null.

In [27]:
df['year'].eq('unknown').sum()

0

In [28]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18626 entries, 62 to 23966
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   date          18626 non-null  object
 1   type          18626 non-null  object
 2   registration  18626 non-null  object
 3   operator      18626 non-null  object
 4   fatalities    18626 non-null  int64 
 5   location      18626 non-null  object
 6   country       18626 non-null  object
 7   cat           18626 non-null  object
 8   year          18626 non-null  object
dtypes: int64(1), object(8)
memory usage: 1.4+ MB


In [29]:
df['country'].value_counts()

USA                     3813
Russia                  1191
Canada                   782
U.K.                     714
Brazil                   566
                        ... 
Mauritius                  1
Christmas Isl.             1
St. Helena                 1
St.Pierre & Miquelon       1
Macau                      1
Name: country, Length: 230, dtype: int64

In [30]:
df['registration'].value_counts()

LZ-...      13
-            9
unreg.       6
N3125N       4
48           4
            ..
44-77538     1
N711TE       1
N326US       1
RP-C82       1
CF-TCU       1
Name: registration, Length: 18228, dtype: int64

In [31]:
df['date'] = pd.to_datetime(df['date'], errors="coerce")

In [32]:
df.isna().sum()

date            171
type              0
registration      0
operator          0
fatalities        0
location          0
country           0
cat               0
year              0
dtype: int64

In [33]:
df.dropna(inplace = True)

In [34]:
df.isna().sum()

date            0
type            0
registration    0
operator        0
fatalities      0
location        0
country         0
cat             0
year            0
dtype: int64

We checked again for the amount values in each column and through data cleaning we went from 23967 values to 18455 values.

In [35]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18455 entries, 62 to 23966
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   date          18455 non-null  datetime64[ns]
 1   type          18455 non-null  object        
 2   registration  18455 non-null  object        
 3   operator      18455 non-null  object        
 4   fatalities    18455 non-null  int64         
 5   location      18455 non-null  object        
 6   country       18455 non-null  object        
 7   cat           18455 non-null  object        
 8   year          18455 non-null  object        
dtypes: datetime64[ns](1), int64(1), object(7)
memory usage: 1.4+ MB


In [36]:
df.head()

date                type registration                operator  \
62 1919-08-11     Felixstowe Fury         N123                     RAF   
63 1920-02-23    Handley Page O/7       G-EANV  Handley Page Transport   
64 1920-02-25  Handley Page O/400       G-EAMC  Handley Page Transport   
65 1920-06-30  Handley Page O/400       G-EAKE  Handley Page Transport   
66 1920-12-14  Handley Page O/400       G-EAMA  Handley Page Transport   

    fatalities             location       country cat  year  
62           1  near Felixtowe RNAS          U.K.  A1  1919  
63           0  Acadia Siding, C...  South Africa  A1  1920  
64           0      near El Shereik         Sudan  A1  1920  
65           0             Östanå        Sweden  A1  1920  
66           4        Golders Green          U.K.  A1  1920

Last but not least we changed the column "year" to int data type using method (".astype")

In [37]:
df['year'] = df['year'].astype(int)

As the last step we converted the the DataFrame to a csv file  using ("to_csv").
We also added ("index = False") to remove the row numbers.

In [38]:
df.to_csv("Aviation-accident-data-test1.csv",index = False)

Below we checked the value counts and the length of each column

In [39]:
df['operator'].value_counts()

USAF                                 930
USAAF                                826
RAF                                  710
US Navy                              469
private                              254
                                    ... 
Serv-Jet                               1
Birgenair, op.for Alas Nacionales      1
Trans National Airlines                1
Million Air San Juan                   1
Private Wings, op.for Alp-Air          1
Name: operator, Length: 5718, dtype: int64

In [40]:
df['type'].value_counts()

Douglas C-47A (DC-3)      1091
Douglas C-47B (DC-3)       363
Douglas C-47 (DC-3)        328
Antonov An-2R              284
DHC-6 Twin Otter 300       249
                          ... 
Douglas DC-6A/C              1
Boeing 727-95                1
CASA C-207C (T.7) Azor       1
Boeing 747-8KZF              1
Boeing 80A                   1
Name: type, Length: 2984, dtype: int64

In [41]:
df.value_counts()

date        type                          registration  operator                 fatalities  location                  country   cat  year
2023-05-15  Learjet 35A                   D-CGFQ        GFD                      2           Hohn Air Base             Germany   A1   2023    1
1965-02-13  Curtiss C-46A                 YS-012C       AESA                     2           Miami, FL                 USA       A1   1965    1
1964-03-01  Lockheed L-049 Constellation  N86504        Paradise Airlines        85          near Lake Tahoe Airpo...  USA       A1   1964    1
1964-03-05  Douglas C-47A (DC-3)          LN-PAS        Widerøes Flyveselskap   0           Oslo-Fornebu Air...       Norway    A1   1964    1
1964-03-08  Douglas C-47A (DC-3)          HK-862        Líneas Aéreas Taxader  28          near Facatavivá          Colombia  A1   1964    1
                                                                                                                                             

In [42]:
df['cat'].value_counts()

A1    13787
A2     2433
O1      854
C1      846
H2      329
C2       65
O2       59
H1       42
U1       25
I2       12
I1        3
Name: cat, dtype: int64

In [43]:
df.head()

date                type registration                operator  \
62 1919-08-11     Felixstowe Fury         N123                     RAF   
63 1920-02-23    Handley Page O/7       G-EANV  Handley Page Transport   
64 1920-02-25  Handley Page O/400       G-EAMC  Handley Page Transport   
65 1920-06-30  Handley Page O/400       G-EAKE  Handley Page Transport   
66 1920-12-14  Handley Page O/400       G-EAMA  Handley Page Transport   

    fatalities             location       country cat  year  
62           1  near Felixtowe RNAS          U.K.  A1  1919  
63           0  Acadia Siding, C...  South Africa  A1  1920  
64           0      near El Shereik         Sudan  A1  1920  
65           0             Östanå        Sweden  A1  1920  
66           4        Golders Green          U.K.  A1  1920

In [44]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18455 entries, 62 to 23966
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   date          18455 non-null  datetime64[ns]
 1   type          18455 non-null  object        
 2   registration  18455 non-null  object        
 3   operator      18455 non-null  object        
 4   fatalities    18455 non-null  int64         
 5   location      18455 non-null  object        
 6   country       18455 non-null  object        
 7   cat           18455 non-null  object        
 8   year          18455 non-null  int32         
dtypes: datetime64[ns](1), int32(1), int64(1), object(6)
memory usage: 1.3+ MB


In [45]:
df['country'].value_counts()

USA               3796
Russia            1177
Canada             779
U.K.               713
Brazil             565
                  ... 
Christmas Isl.       1
Mauritius            1
Cook Isl.            1
Tuvalu               1
Monaco               1
Name: country, Length: 230, dtype: int64